### PR TITLE
Fix index out of bounds Exception in ProgrammingSubmissionService

### PR DESCRIPTION
**Pull request by SimonEntholzer**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I documented the Java code using JavaDoc style.

### Motivation and Context

In some case a submission for a programming exercise might not have a result, even no automatic result. If this event comes to pass, which we did not consider before, the `lockSubmission()` method in the `ProgrammingSubmissionService` tries to access the submissions' result with id `-1`. 

**EDIT:** The issue can only happen with older programming exercises. Before the addition of multiple results, the new manual result was in a separate submission. Therefore, the sumbission only had one result when it got loaded from the databse. If the assessment of such submissions is canceled now, it leaves behind a submission without any results, which leads to the mentioned problem.  With newly created Programming Exercises, this issue will not happen, as there will always be an automatic result.(Unless it is deleted manually, and this case is covered now, too.)

<details>
<summary>Error log</summary>
<p>

```c#
Jan 30 11:15:20 vmbruegge105 java[27609]: java.lang.IndexOutOfBoundsException: Index -1 out of bounds for length 0
Jan 30 11:15:20 vmbruegge105 java[27609]:         at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:248)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at java.base/java.util.Objects.checkIndex(Objects.java:359)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at java.base/java.util.ArrayList.get(ArrayList.java:427)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at de.tum.in.www1.artemis.domain.Submission.getResultForCorrectionRound(Submission.java:127)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at de.tum.in.www1.artemis.service.ProgrammingSubmissionService.lockSubmission(ProgrammingSubmissionService.java:701)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at de.tum.in.www1.artemis.service.ProgrammingSubmissionService.lockAndGetProgrammingSubmissionWithoutResult(ProgrammingSubmissionService.java:679)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at de.tum.in.www1.artemis.service.ProgrammingSubmissionService$$FastClassBySpringCGLIB$$fbbf5f1c.invoke(<generated>)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:688)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at de.tum.in.www1.artemis.service.ProgrammingSubmissionService$$EnhancerBySpringCGLIB$$6fd5313b.lockAndGetProgrammingSubmissionWithoutResult(<generated>)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at de.tum.in.www1.artemis.web.rest.ProgrammingSubmissionResource.getProgrammingSubmissionWithoutAssessment(ProgrammingSubmissionResource.java:435)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at de.tum.in.www1.artemis.web.rest.ProgrammingSubmissionResource$$FastClassBySpringCGLIB$$9996c806.invoke(<generated>)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:779)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:750)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at org.springframework.security.access.intercept.aopalliance.MethodSecurityInterceptor.invoke(MethodSecurityInterceptor.java:61)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:750)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:97)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:750)
```

</p>
</details>  

### Description
We added a check to ensure that a submission without any results is handled differently from usual submissions. When assessing it new feedback is created, but it is also clear that no automatic tests ran.
image

### Steps for Testing

In the normal case there is either _no submission_ or a _submission that has an automatic result_. (_Unless I really did miss something_) I guess what theoretically could have happened is, that a submission was created, but somehow there was an error somewhere else on bitbucket/bamboo, and no result was received and saved. 
You can either test this in exam-mode or in normal exercise mode, in both I couldn't reproduce the initial problem naturally.
<details>
<summary>Things I have tried to reproduce it naturally:</summary>

- submit the exam but not the exercise
- submit the exercise, but not the exam
- after the exam is over use the "assess unsubmitted exams" feature
- under participations use the "trigger" feature (if the student has used the submit button there is no submission)
- different ways to access the code/assessment editor: "new assessment" from the `exercise-assessment-dashboard` or "next assessment" in the `code-editor`

</details>

What you can do is manipulate the database by hand to get in the situation that causes the error.

1. create an programming exrcise / an exam with an programming exercise
2. participate with a student, submit
3. in MySQL, remove the automatically created result + feedbacks of this submission from your database
4. now you can go to the assessment dashboard -> new assessment
image
5. you will notice that the test part of the resultString is undefined, but the code-editor opens.

On develop the new assessment action leads to the out of bounds exception, and therefore doesn't open the code-editor.

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java [hunk 1]
```diff
@@ -657,10 +657,10 @@ public ProgrammingSubmission findByIdWithEagerResultAndFeedback(long submissionI
      * @param submissionId the id of the programming submission
      * @return the locked programming submission
      */
-    public ProgrammingSubmission lockAndGetProgrammingSubmission(Long submissionId) {
+    public ProgrammingSubmission lockAndGetProgrammingSubmission(Long submissionId, int correctionRound) {
```
Comment (FrankeLukas) [inline src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java:660]: The documentation is missing for the new parameter.

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java [hunk 2]
```diff
@@ -657,10 +657,10 @@ public ProgrammingSubmission findByIdWithEagerResultAndFeedback(long submissionI
      * @param submissionId the id of the programming submission
```
Comment (SimonEntholzer) [inline src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java:657]:
```suggestion
     * @param submissionId the id of the programming submission
     * @param correctionRound the correctionRound of the programming submission
```

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java (hunk 1 above)
Comment (SimonEntholzer) [reply src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java:661]: I added it :)

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java [hunk 3]
```diff
@@ -697,11 +698,17 @@ protected Result lockSubmission(Submission submission, int correctionRound) {
         if (correctionRound == 0 && submission.getLatestResult() != null && AssessmentType.AUTOMATIC.equals(submission.getLatestResult().getAssessmentType())) {
             existingResult = submission.getLatestResult();
         }
+        else if (correctionRound == 0 && submission.getLatestResult() == null) {
+            // if the student did not participate in the exam, and therefore has no result
```
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java:702]: It is a little bit confusing that you are talking about exams suddenly here. Maybe write something like: `This edge case can happen in the exam mode if the student did not participate ......`

**Review by stefanwaldhauser — CHANGES_REQUESTED** [review]
image
Everything worked as described! Gj. I manually deleted the result and feedback in the database. I could still open the submission and give feedback. Just very small things in the code :)

**Review by airelawaleria — APPROVED** [review]
Me and @JohannesWeiss could have reproduced the issue with old programming exercises. When trying to "start new assessment" from assessment dashboard, cancelling and starting again multiple times, it leads to index out of bound exception - 1. 

The isssue seems to be solved.

**Review by JohannesWeiss — DISMISSED** [review]
Tested with @airelawaleria. The issue seems to be resolved, could not reproduce the error with an exercise where it occurred previously.

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java (hunk 3 above)
Comment (SimonEntholzer) [reply src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java:702]: I changed the comment, as it turns out that wasn't the origin of the issue

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java [hunk 4]
```diff
@@ -697,11 +698,17 @@ protected Result lockSubmission(Submission submission, int correctionRound) {
...
+            existingResult = null;
```
[the remark this answers is not in the document]
Comment (SimonEntholzer) [reply src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java]: I changed it, indeed that looks cleaner :)

**Review by airelawaleria — APPROVED** [review]

**Review by JohannesWeiss — APPROVED** [review]

**Review by EneaGore — APPROVED** [review]

**Review by krusche — APPROVED** [review]

**Pull request merged**